File: tests/real-world/b03_stack.go
package main

import (
	"errors"
	"fmt"
)

type Stack struct {
	items []int
}

func NewStack() *Stack {
	return &Stack{items: make([]int, 0)}
}

func (s *Stack) Push(val int) {
	s.items = append(s.items, val)
}

func (s *Stack) Pop() (int, error) {
	if len(s.items) == 0 {
		return 0, errors.New("stack is empty")
	}
	val := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]
	return val, nil
}

func (s *Stack) Peek() (int, error) {
	if len(s.items) == 0 {
		return 0, errors.New("stack is empty")
	}
	return s.items[len(s.items)-1], nil
}

func (s *Stack) Size() int {
	return len(s.items)
}

func (s *Stack) IsEmpty() bool {
	return len(s.items) == 0
}

func main() {
	s := NewStack()
	s.Push(10)
	s.Push(20)
	s.Push(30)
	fmt.Printf("Size: %d\n", s.Size())

	val, err := s.Pop()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	fmt.Printf("Popped: %d\n", val)

	top, err := s.Peek()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	fmt.Printf("Top: %d\n", top)
	fmt.Printf("Empty: %v\n", s.IsEmpty())
}

File: tests/java-real-world/b03_stack.java
import java.util.ArrayList;
import java.util.EmptyStackException;

public class b03_stack {

    static class GenericStack<T> {
        private final ArrayList<T> items = new ArrayList<>();

        void push(T item) { items.add(item); }

        T pop() {
            if (isEmpty()) throw new EmptyStackException();
            return items.remove(items.size() - 1);
        }

        T peek() {
            if (isEmpty()) throw new EmptyStackException();
            return items.get(items.size() - 1);
        }

        boolean isEmpty() { return items.isEmpty(); }
        int size() { return items.size(); }
    }

    public static void main(String[] args) {
        GenericStack<Integer> stack = new GenericStack<>();
        System.out.println("Empty: " + (stack.isEmpty() ? "PASS" : "FAIL"));

        stack.push(10); stack.push(20); stack.push(30);
        System.out.println("Size after 3 pushes: " + (stack.size() == 3 ? "PASS" : "FAIL"));
        System.out.println("Peek: " + (stack.peek() == 30 ? "PASS" : "FAIL"));
        System.out.println("Pop: " + (stack.pop() == 30 ? "PASS" : "FAIL"));
        System.out.println("Pop: " + (stack.pop() == 20 ? "PASS" : "FAIL"));
        System.out.println("Pop: " + (stack.pop() == 10 ? "PASS" : "FAIL"));
        System.out.println("Empty after pops: " + (stack.isEmpty() ? "PASS" : "FAIL"));

        try { stack.pop(); System.out.println("Pop empty: FAIL"); }
        catch (EmptyStackException e) { System.out.println("Pop empty throws: PASS"); }

        // Test with String type
        GenericStack<String> sStack = new GenericStack<>();
        sStack.push("hello"); sStack.push("world");
        System.out.println("String peek: " + ("world".equals(sStack.peek()) ? "PASS" : "FAIL"));
    }
}
